docs/src/modules/ROOT/pages/using-timefold-solver/running-the-solver.adoc

```diff
@@ -102,6 +102,59 @@ This mode is reproducible (see the reproducible mode). It is also intrusive beca
 
 The FULL_ASSERT mode is horribly slow (because it does not rely on incremental score calculation).
 
+[#environmentModeTracedFullAssert]
+=== `TRACED_FULL_ASSERT`
+
+The TRACED_FULL_ASSERT mode turns on all the <<environmentModeFullAssert, FULL_ASSERT>> assertions and additionally track changes to the working solution.
+This allows the solver to report exactly what variables were corrupted and what variable listener events are missing.
+
+For instance, when this incorrect move is evaluated
+
+```java
+    @Override
+    public Move<Solution_> createUndoMove(ScoreDirector<Solution_> scoreDirector) {
+        return this;
+    }
+
+    @Override
+    protected void doMoveOnGenuineVariables(ScoreDirector<Solution_> scoreDirector) {
+        scoreDirector.beforeVariableChanged(entity, "value");
+        entity.setValue(newValue);
+        scoreDirector.afterVariableChanged(entity, "value");
+    }
+```
+
+the exception thrown by the solver will have the following in its message:
+
+```
+Variables that are different between before and undo:
+  - Actual value (newValue) of variable value on EntityClass entity (entity) differs from expected (oldValue)
+```
+
+In particular, the solver will recalculate all shadow variables from scratch on the solution after the undo and then report:
+
+- Genuine and shadow variables that are different between before and undo.
+After an undo move is evaluated, it is expected to exactly match the original working solution.
+
+- Variables that are different between from scratch and before.
+This is to detect if the solution was corrupted before the move was evaluated due to shadow variable corruption.
+
+- Variables that are different between from scratch and undo.
+This is to detect if recalculating the shadow variables from scratch is different from the incremental shadow variable calculation.
+
+- Missing variable listener events for actual move.
+Any variable that changed between the before move solution and after move solution without either a
+`beforeVariableChanged` or `afterVariableChanged` would be reported here.
+
+- Missing variable listener events for undo move.
+Any variable that changed between the after move solution and after undo move solution without either a
+`beforeVariableChanged` or `afterVariableChanged` would be reported here.
+
+This mode is reproducible (see the reproducible mode).
```

Comment:
The reproducible mode deserves a direct reference. (Unless it doesn't have an ID.)